### PR TITLE
Text formatter to convert a sentence to LaTeX format

### PR DESCRIPTION
**Pull request by hollyyuqizheng**
I added a "Format" button to each sentence on the story's page. This button appears for the sentence that the user's mouse is currently hovering over. 
When clicked, this button will trigger a conversion between this sentence to its LaTeX format. A popup window will show up with some metadata as well as the formatted LaTeX code. 

Screen Shot 2021-03-20 at 16 02 46

Screen Shot 2021-03-20 at 15 47 37

### CODE REVIEW

File: jsx/App/Stories/Story/Display/TextFormatButton.jsx [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        let dependents;
+        if (! sentences[0]) {
+            dependents = sentences["dependents"]
+        } else {
+            dependents = sentences[0]["dependents"];
+        }
+
+        const aingaeWordList = dependents[0]["values"];
+        const morphemeList = dependents[1]["values"];
+        const glossList = dependents[3]["values"];
```
Comment (sciepsilon) [inline jsx/App/Stories/Story/Display/TextFormatButton.jsx:19]: A lot of LingView's value comes from its flexibility in supporting arbitrary combinations of tiers. This code assumes that the vernacular words are in dependent tier 0, morphemes are 1, and glosses are 3. That assumption can be wrong. 

For example it would be reasonable for an ELAN file to have the following 8 tiers, in order: 
- narrow transcription
- broad transcription
- morphemes
- morpheme translation into Spanish
- morpheme translation into English
- free translation into Spanish
- free translation into English
- suprasegmental features (e.g. creak, falsetto)

Content curators can set the tier order in ELAN files to be anything they like. FLEx files are more consistent than ELAN, but I expect variation there too. 

In a perfect world, LingView would automatically generate good LaTeX glosses regardless of what tiers are present - but in practice, I don't see a way to do this 100% of the time. So, maybe we need user input. When someone clicks the format button, we can ask them to select which tiers to use. Or, LingView can guess, but we let the user change tiers afterwards when they're viewing the LaTeX format. 

We can consider leaving the assumptions built-in, with the option to add flexibility later. If we do this, I'd like to a) make sure we're using the right format for our corpora, and b) put a note on the output page encouraging users to swap out the tiers if needed.

File: jsx/App/Stories/Story/Display/TextFormatButton.jsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            "gloss" : word2Gloss
+        };
+    }
+
+    /* Retrieves the English translation for the sentence. */
+    function getSentenceTranslation(dependents) {
+        // Find the tier that corresponds to the English translation. 
+        for (let i = 0; i < dependents.length; i+=1) {
+            const tierName = dependents[i]["tier"].toLowerCase();
+            if (tierName == "frase en inglés" || tierName == "english") {
```
Comment (sciepsilon) [inline jsx/App/Stories/Story/Display/TextFormatButton.jsx:112]: The glossing language isn't always English. (In our corpora, it's always English and/or Spanish, but LingView can also be used with other languages.) Even when it's English, the gloss tier can be named anything. If the file comes from ELAN, we use whatever tier names were in the original document.

File: jsx/App/Stories/Story/Display/TextFormatButton.jsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            title : title,
+            sentenceUrl : sentenceUrl,
+            morphemes : morphemeMap,
+            gloss : glossMap,
+            sentenceTranslation : sentenceTranslation  
+        };
+    }
+
+    /* Returns a map between each word and all of its sub-components (core and clitics, and gloss, etc.) */
+    function organizeWords(wordList, morphemeList, glossList) {
```
Comment (sciepsilon) [inline jsx/App/Stories/Story/Display/TextFormatButton.jsx:42]: A lot of this logic belongs in preprocessing, not in jsx. Preprocessing is when we have the most information about the original FLEx or ELAN file. Also, it's better for preprocessing to be slow than for the jsx to be slow. Preprocessing only happens once per update, whereas jsx happens every time a user loads the page. 

Adding to the preprocessing logic means we'll need to change or add to the json format that LingView uses to store files. All changes to LingView's json format need to be backwards-compatible, so I'd like to talk over the proposed changes before you put too much effort into coding and certainly before we merge them into BrownCLPS master.

**Review by sciepsilon — CHANGES_REQUESTED** [review]
Needs a substantial redesign in order to work well with FLEx and ELAN files in the full variety of languages and formats. Might be a good idea to for us to discuss the design before you implement it. I also recommend testing on a variety of FLEx and ELAN files, including some from Brown's corpora and some from other corpora or that you create yourself to test edge cases.

**Comment by hollyyuqizheng** [comment]
Will close this PR for now
